### PR TITLE
Annotation merge

### LINKED ISSUE #430: Javamerge does not merge annotations
Currently, the java merger does not merge annotations.
So far there are two restriction after implementation of this ticket:
1. Annotations are not merged recursively as it's not easy to identify "matching" annotations within an array
2. [Repeating annotations](https://docs.oracle.com/javase/tutorial/java/annotations/repeating.html) defined next to each other could not be identified / differentiated for a proper merge, therefore, we need to overwrite / ignore all annotations of a given scope as a block

*Acceptance Criteria*
* [x] Annotations, which are not yet there should be added
* [x] All annotations should be replaced or not replaced
* [x] Annotations should be merged the same for fields, methods, classes

### PR DESCRIPTION
**Pull request by MikeSchumacherCapgemini**
Adresses/Fixes #430.

Implements

* Update to latest qdox version from version M2 to M10
* Added new modifyable classes to override default implementation

### CODE REVIEW

**Comment by MikeSchumacherCapgemini** [comment]
@maybeec

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableJavaClass.java [hunk 1]
```diff
@@ -291,59 +302,59 @@ public String getGenericValue() {
...
+    //
+    // @Override
+    // public String getClassNamePrefix() {
+    // return getFullyQualifiedName() + "$";
+    // }
+    //
+    // // @Override
+    // // public JavaType asType() {
+    // // return this;
+    // // }
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableJavaClass.java]: Is this code entirely not needed anymore? Then we should remove it.

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableJavaClass.java [hunk 2]
```diff
@@ -552,29 +563,29 @@ public void addClass(JavaClass cls) {
...
-        return getNestedClasses();
-    }
+    // /**
+    // * @deprecated Use {@link #getNestedClasses()} instead.
+    // */
+    // @Override
+    // @Deprecated
+    // public List<JavaClass> getClasses() {
+    // return getNestedClasses();
+    // }
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableJavaClass.java]: same here.

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableJavaClass.java [hunk 3]
```diff
@@ -552,29 +563,29 @@ public void addClass(JavaClass cls) {
...
-    }
+    // @Override
+    // public JavaClass getInnerClassByName(String name) {
+    // return getNestedClassByName(name);
+    // }
+    //
+    // @Override
+    // public List<JavaClass> getInnerClasses() {
+    // return getNestedClasses();
+    // }
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableJavaClass.java]: same here.

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableJavaClass.java [hunk 4]
```diff
@@ -703,10 +714,10 @@ public BeanProperty getBeanProperty(String propertyName, boolean superclasses) {
...
     }
 
-    @Override
-    public JavaClass getDeclaringClass() {
-        return getParentClass();
-    }
+    // @Override
+    // public JavaClass getDeclaringClass() {
+    // return getParentClass();// getSuperJavaClass();
+    // }
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableJavaClass.java]: same here.

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 5]
```diff
@@ -185,17 +210,27 @@ public void beginClass(ClassDef def) {
...
+        // classStack.getFirst().addClass(newClass);
+        // currentField.setEnumConstantClass(newClass);
+        // } else if (!classStack.isEmpty()) {
+        // classStack.getFirst().addClass(newClass);
+        // // newClass.setSuperClass(classStack.getFirst());
+        // newClass.setParentClass(classStack.getFirst());
+        // } else {
+        // source.addClass(newClass);
+        // }
+        // return newClass;
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: same here.

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 6]
```diff
@@ -367,15 +421,27 @@ public void endMethod(MethodDef def) {
...
 
-        currentField = new DefaultJavaField();
-        currentField.setParentClass(classStack.getFirst());
+        currentField = new DefaultJavaField(def.getName());
+        currentField.setDeclaringClass(classStack.getFirst());// .setParentClass(classStack.getFirst());
         currentField.setLineNumber(def.getLineNumber());
         currentField.setModelWriterFactory(modelWriterFactory);
 
-        currentField.setName(def.getName());
+        // currentField.setName(def.getName());
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: same here.

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 7]
```diff
@@ -367,15 +421,27 @@ public void endMethod(MethodDef def) {
...
+        return result;
+    }
+
     @Override
     public void beginField(FieldDef def) {
 
-        currentField = new DefaultJavaField();
-        currentField.setParentClass(classStack.getFirst());
+        currentField = new DefaultJavaField(def.getName());
+        currentField.setDeclaringClass(classStack.getFirst());// .setParentClass(classStack.getFirst());
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: same here.

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 8]
```diff
@@ -185,17 +210,27 @@ public void beginClass(ClassDef def) {
...
 
     protected ModifyableJavaClass bindClass(ModifyableJavaClass newClass) {
-
         if (currentField != null) {
             classStack.getFirst().addClass(newClass);
             currentField.setEnumConstantClass(newClass);
         } else if (!classStack.isEmpty()) {
             classStack.getFirst().addClass(newClass);
-            newClass.setParentClass(classStack.getFirst());
+            newClass.setDeclaringClass(classStack.getFirst());
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: did you compare and change such changes based on the DefaultModelBuilder?

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 9]
```diff
@@ -348,17 +390,29 @@ public void endMethod(MethodDef def) {
...
+
+        TypeResolver typeResolver =
+            TypeResolver.byClassName(declaringClass.getBinaryName(), classLibrary, source.getImports());
+
+        ModifyableJavaTypeVariable<G> result =
+            new ModifyableJavaTypeVariable<G>(typeVariableDef.getName(), typeResolver);
 
         if (typeVariableDef.getBounds() != null && !typeVariableDef.getBounds().isEmpty()) {
-            List<JavaType> bounds = new LinkedList<>();
+            List<JavaType> bounds = new LinkedList<JavaType>();
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: the type in the constructor is not needed, it even throws a warning, that it is redundant. You simply can use <>.

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 10]
```diff
@@ -397,12 +463,20 @@ public void beginField(FieldDef def) {
...
+                    TypeResolver.byClassName(classStack.getFirst().getBinaryName(), classLibrary, source.getImports());
+            }
+
             // DefaultExpressionTransformer??
-            DefaultJavaAnnotationAssembler assembler = new DefaultJavaAnnotationAssembler(currentField);
+            DefaultJavaAnnotationAssembler assembler =
+                new DefaultJavaAnnotationAssembler(currentField.getDeclaringClass(), classLibrary, typeResolver);
 
-            List<Expression> arguments = new LinkedList<>();
+            List<Expression> arguments = new LinkedList<Expression>();
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: unecessary type declaration

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 11]
```diff
@@ -422,9 +496,11 @@ public void addParameter(FieldDef fieldDef) {
             new ExtendedJavaParameter(createType(fieldDef.getType(), fieldDef.getDimensions()), fieldDef.getName(),
                 fieldDef.getModifiers(), fieldDef.isVarArgs());
         if (currentMethod != null) {
-            jParam.setDeclarator(currentMethod);
+            jParam.setExecutable(currentMethod);
+            // .setDeclarator(currentMethod);
         } else {
-            jParam.setDeclarator(currentConstructor);
+            jParam.setExecutable(currentMethod);
+            // .setDeclarator(currentConstructor);
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: remove commented code

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 12]
```diff
@@ -422,9 +496,11 @@ public void addParameter(FieldDef fieldDef) {
             new ExtendedJavaParameter(createType(fieldDef.getType(), fieldDef.getDimensions()), fieldDef.getName(),
                 fieldDef.getModifiers(), fieldDef.isVarArgs());
         if (currentMethod != null) {
-            jParam.setDeclarator(currentMethod);
+            jParam.setExecutable(currentMethod);
+            // .setDeclarator(currentMethod);
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: remove commented code

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java [hunk 13]
```diff
@@ -433,12 +509,19 @@ public void addParameter(FieldDef fieldDef) {
...
+                typeResolver =
+                    TypeResolver.byClassName(classStack.getFirst().getBinaryName(), classLibrary, source.getImports());
+            }
+
             DefaultJavaAnnotationAssembler assembler =
-                new DefaultJavaAnnotationAssembler((JavaAnnotatedElement) entity);
+                new DefaultJavaAnnotationAssembler(entity.getDeclaringClass(), classLibrary, typeResolver);
 
-            List<JavaAnnotation> annotations = new LinkedList<>();
+            List<JavaAnnotation> annotations = new LinkedList<JavaAnnotation>();
```
Comment (maybeec) [inline cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: unneccessary type declaration in constructor

**Comment by maybeec** [comment]
@MikeSchumacherCapgemini could you shortly answer the open questions in my comments?

File: cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java (hunk 8 above)
Comment (MikeSchumacherCapgemini) [reply cobigen/cobigen-javaplugin-parent/cobigen-javaplugin/src/main/java/com/devonfw/cobigen/javaplugin/merger/libextension/ModifyableModelBuilder.java]: setDeclaringClass seems to simply be the renamed setParentClass.
In general, the term "parent" seems to have been replaced by "declaring".

**Pull request merged**